src/common/libutil/grudgeset.c

```diff
@@ -0,0 +1,134 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <string.h>
+#include <errno.h>
+#include <czmq.h>
+
+#include <jansson.h>
+
+struct grudgeset {
+    json_t *set;
+    zhashx_t *grudges;
+};
+
+void grudgeset_destroy (struct grudgeset *gset)
+{
+    if (gset) {
+        int saved_errno = errno;
+        json_decref (gset->set);
+        zhashx_destroy (&gset->grudges);
+        free (gset);
+        errno = saved_errno;
+    }
+}
+
+static struct grudgeset *grudgeset_create (void)
+{
+    struct grudgeset *gset = calloc (1, sizeof (*gset));
+    if (!gset
+        || !(gset->set = json_array ())
+        || !(gset->grudges = zhashx_new ()))
+        goto error;
+    return gset;
+error:
+    grudgeset_destroy (gset);
+    return NULL;
+}
+
+int grudgeset_used (struct grudgeset *gset, const char *val)
+{
+    if (!gset || !val)
+        return 0;
+    return (zhashx_lookup (gset->grudges, val) != NULL);
+}
+
+static int json_array_find (json_t *array, const char *val)
+{
+    size_t index;
+    json_t *entry;
+    json_array_foreach (array, index, entry) {
+        const char *s = json_string_value (entry);
+        if (val && strcmp (val, s) == 0) {
+            return (int) index;
+        }
+    }
+    return -1;
+}
+
+int grudgeset_contains (struct grudgeset *gset, const char *val)
+{
+    if (!gset
+        || !val
+        || json_array_find (gset->set, val) < 0)
+        return 0;
+    return 1;
+}
+
+int grudgeset_append (struct grudgeset **gsetp, const char *val)
+{
+    json_t *o = NULL;
+
+    if (!gsetp || !val) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (!*gsetp) {
+        if (!(*gsetp = grudgeset_create ()))
+            return -1;
+    }
+    if (grudgeset_used (*gsetp, val)) {
+        errno = EEXIST;
+        return -1;
+    }
+    if (!(o = json_string (val))
+        || json_array_append_new ((*gsetp)->set, o) < 0) {
+        json_decref (o);
+        errno = ENOMEM;
+        return -1;
+    }
```

Comment:
It might simplify the json code a bit if an object were used instead of an array.

It wouldn't be necessary to do the linear search to test or remove, and you could use a non-allocated value like `json_null()` when inserting.

Reply:
It is non-obvious,  but the reason an array is used here is so that the current set can be used without copying in the response to an RPC querying the current set. Maybe that was a poor optimization?

Reply:
Ah I saw the `_tojson` comment but failed to think that through.

Since it matters to users, maybe the header should mention that it returns a json array?

Reply:
Uh, yeah, good thought :facepalm: